lib/lexer/lexer.cpp

```diff
@@ -0,0 +1,484 @@
+	}
+	state_data.pushState(state);
+	return current;
+}
+
+
+Lexer::State Lexer::parameterExpansionState() {
+	return generalExpansionHandler(state, '}');
+}
+
+Lexer::State Lexer::commandSubstitutionState() {
+	return generalExpansionHandler(state, ')');
+}
+
+// Requires some extra logic due to having multiple terminating characters
+Lexer::State Lexer::arithmeticExpansionState() {
+	auto state = State::ArithmeticExpansion;
+	while (state == State::ArithmeticExpansion) {
+		char ch = chars.consume();
+		if (ch == EOF) {
+			return state_data.popState();
+		}
+		state_data.word.push_back(ch);
+		if (ch == ')') {
+			if (chars.peek() == ')') {
+				state_data.word.push_back(chars.consume());
+				return state_data.popState();
+			} else {
+				// If it closes with ')' instead of '))', we are actually in a subshell of a command substitution
+				// Example: $((echo ...) | cat)
+				return State::CommandSubstitution;
+			}
+		}
+		state = nextNestedState(State::ArithmeticExpansion, ch);
+	}
+	state_data.pushState(State::ArithmeticExpansion);
+	return state;
+}
+
+Lexer::State Lexer::backQuoteState() {
+	while (true) {
+		char ch = chars.consume();
+		state_data.word.push_back(ch);
+		if (isBackslash(ch)) {
+			state_data.pushState(State::BackQuote);
+			return State::InnerBackslash;
+		} else if (isBackQuote(ch)) {
+			return state_data.popState();
+		}
+	}
+}
+
+/* State Data methods */
+
+void Lexer::StateData::pushState(State state) {
+	states.push(state);
+}
+
+Lexer::State Lexer::StateData::popState() {
+	D_ASSERT(!states.empty());
+	State state = states.top();
+	states.pop();
+	return state;
+}
+
+Lexer::State Lexer::StateData::previousState() {
+	D_ASSERT(!states.empty());
+	return states.top();
+}
+
+// https://www.gnu.org/software/bash/manual/bash.html#index-metacharacter
+#define WHITESPACE                " \t\n"
+#define OPERATOR_START_CHARACTERS "|&;()<>"                            // characters that operators can start with
+#define OPERATOR_CHARACTERS       OPERATOR_START_CHARACTERS "-"        // characters that _can_ appear in operators
+#define METACHARACTERS            WHITESPACE OPERATOR_START_CHARACTERS // characters that _always_ delimit words
+
+static bool contains(const char *s, char ch) {
+	return strchr(s, ch) != nullptr;
+}
+
+bool isCommentStart(char ch) {
+	return ch == '#';
+}
+
+bool isSpace(char ch) {
+	return contains(WHITESPACE, ch);
+}
+
+bool isMetaCharacter(char ch) {
+	return contains(METACHARACTERS, ch);
+}
+
+bool isOperatorCharacter(char ch) {
+	return contains(OPERATOR_CHARACTERS, ch);
+}
+
+bool isOperatorStart(char ch) {
+	return contains(OPERATOR_START_CHARACTERS, ch);
+}
+
+bool isNewline(char ch) {
+	return ch == '\n';
+}
+
+bool isBackslash(char ch) {
+	return ch == '\\';
+}
+
+bool isSingleQuote(char ch) {
+	return ch == '\'';
+}
+
+bool isDoubleQuote(char ch) {
+	return ch == '\"';
+}
+
+bool isDollarSign(char ch) {
+	return ch == '$';
+}
+
+bool isBackQuote(char ch) {
```

Comment:
`isBackTick` ?

Reply:
I changed it to backquote: https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap03.html#tag_03_35

Reply:
Ah interesting, LGTM 👍